lib/http2.c

```diff
@@ -1619,8 +1619,15 @@ static ssize_t http2_recv(struct Curl_easy *data, int sockindex,
     return -1;
   }
 
-  if(stream->closed)
-    /* closed overrides paused */
+  /*
+   * Closed overrides paused.
+   * However the residual data right after finished protocol switching (h2c
+   * upgrade) needs special handling, for some servers (e.g. nghttp2's reverse
+   * proxy) may start sending the response payload immediately following the
+   * server-side connection preface, other than waiting for the client's
+   * connection preface and possible setting negotiations.
+   */
+  if(stream->closed && (data->req.upgr101 != UPGR101_RECEIVED))
```

Comment:
The condition for continuing isn't really that it got a 101 back though is it? Isn't it rather that there are more bytes queued for this stream to process?

Reply:
Hi @bagder , thanks for the review! I have to apologize in advance for my lack of expertise of this project, and thus the possibility that I'm having incorrect assumptions or giving improper suggestions.

Based on my current understanding, the major problem we're facing here is that when `curl` handles the protocol switching for HTTP/2 and submits the payload for `nghttp2` to parse, there is a chance that stream 1 is fulfilled and closed right away if the server sends it back too early:
https://github.com/curl/curl/blob/1e19eceb50d6d8b9c9b37e2f8e1259b2f311c82a/lib/http2.c#L2327-L2339

Once that happens, the residual payload (parsed by `nghttp2` and pushed to the stream buffer by the callbacks) would still await retrieval while being marked as dirty (`data->state.drain` being non-zero).

Also such payload gets ignored by curl due to the new shortcut added in 252790c53:
https://github.com/curl/curl/blob/1e19eceb50d6d8b9c9b37e2f8e1259b2f311c82a/lib/http2.c#L1622-L1624

My motivation is therefore to allow stream 1 to go beyond this shortcut and reach the draining branch:
https://github.com/curl/curl/blob/1e19eceb50d6d8b9c9b37e2f8e1259b2f311c82a/lib/http2.c#L1658-L1661

Therefore the payload may be picked up at a later time at:
https://github.com/curl/curl/blob/1e19eceb50d6d8b9c9b37e2f8e1259b2f311c82a/lib/http2.c#L1780-L1791

> Isn't it rather that there are more bytes queued for this stream to process?

This is because I wanted to avoid altering too much of the logic (accidentally covering other cases) and the only case I know of that shall be allowed to go beyond the shortcut is upon `UPGR101_RECEIVED`.

Actually I wanted to also include `stream->stream_id == 1` to the condition, but that cannot be done because the stream ID would have been destroyed already as part of the `on_stream_close` callback.

I have to say that I myself am not too happy with such a solution. I may try to think of other possible approaches.